Clean closures transitively.

andrewsmartin opened the pull request:
(Adapted from https://github.com/spotify/scio/pull/698)

`ClosureCleaner` seems to not work as well as it could in certain cases, especially when it comes to multiple layers of nesting. For example it's possible to have a closure whose enclosing scope references an outer that is not serializable, but such that (by tracing the transitive method calls / field references starting from the inner closure), we don't actually need the reference, so we can null it out. The test case I added to ClosureTest is an example of this.

andrewsmartin committed:
Clean closures transitively.

johnynek commented on chill-scala/src/main/scala/com/twitter/chill/ClosureCleaner.scala:
```diff
@@ -32,233 +32,205 @@ package com.twitter.chill
-  private val innerClasses = MMap[Class[_], Set[Class[_]]]()
-  private val accessedFieldsMap = MMap[Class[_], Set[Field]]()
+trait ClosureCleaner {
+  val OUTER = "$outer"
```
can we make this `final`?

johnynek commented on chill-scala/src/main/scala/com/twitter/chill/ClosureCleaner.scala:
```diff
@@ -32,233 +32,205 @@ package com.twitter.chill
-    outerFields
-      .getOrElseUpdate(c,
-        c.getDeclaredFields.find { _.getName == OUTER })
+  def copyField(f: Field, old: AnyRef, newv: AnyRef) {
```
can we add `: Unit = ` since procedure syntax is deprecated?

johnynek commented on chill-scala/src/main/scala/com/twitter/chill/ClosureCleaner.scala:
```diff
@@ -32,233 +32,205 @@ package com.twitter.chill
+
+  def outerFieldOf(c: Class[_]): Option[Field] = Try(c.getDeclaredField(OUTER)).toOption
+
+  def setOuter(obj: AnyRef, outer: AnyRef) {
```
`:Unit =`

johnynek commented on chill-scala/src/main/scala/com/twitter/chill/ClosureCleaner.scala:
```diff
@@ -32,233 +32,205 @@ package com.twitter.chill
-        toF(s)
+  private def innerClasses: Set[Class[_]] = {
+    val seen = MSet[Class[_]](funcClass)
+    var stack = MStack[Class[_]](funcClass)
```
we don't need a `var` to a mutable stack do we? I think a `val` here is fine and no need to reassign below (or use a `var` to a `List[Class[_]]` as the stack, or use a tail-recursive loop and no var or mutable. :)

johnynek commented on chill-scala/src/main/scala/com/twitter/chill/ClosureCleaner.scala:
```diff
@@ -32,233 +32,205 @@ package com.twitter.chill
   }
 }
 
-class InnerClosureFinder(output: MSet[Class[_]]) extends ClassVisitor(ASM5) {
```
this looks like you've just reformatted, right?

johnynek commented on chill-scala/src/test/scala/com/twitter/chill/ClosureCleanerSpec.scala:
```diff
@@ -67,5 +70,29 @@ class ClosureCleanerSpec extends WordSpec with Matchers {
+
+      val fn = new NestedClosuresNotSerializable().getMapFn
+      ClosureCleaner(fn)
+      isSerializable(fn) shouldBe true
```
is this false before? Can we add a test that checks that?

Ideally we see something fail until we run the cleaner.

johnynek reviewed (commented):
thanks for the PR.

had a couple of minor comments and a request for a bit more testing (I want us to assert that the test wouldn't pass without the clean call, else we don't know the test is doing anything).

Can you comment and highlight where the diff is? It's a bit hard to follow since there are so many cosmetic changes to the code mixed in.

andrewsmartin commented:
Sure, I will add some extra asserts showing that the closure is not serializable before cleaning. Also sorry about the mess of changes, I was initially thinking of just patching this in Scio and pretty much just copied it back here. I'll try to tidy it up a bit so it's easier to follow.

The major difference here is using the `accessedFields` map to null any outers out if they are not transitively referenced by the starting closure. The test illustrates how this is possible, because although the closure returned from `NestedClosuresNotSerializable.getMapFn` is enclosed by a scope which has the "not-serializable" object as its outer, it can still be made serializable since the inner closure does not actually reference it via any of its method calls. In other words, the path by which the intermediate scope does reference it is not reachable by any method path from the inner closure. In this test the body of `val fn = { }` is enclosed by the body of `closure("one")`, which in turn is enclosed by `SomethingNotSerializable`. Here we can set the outer of `closure("one")` to null (instead of the not-serializable object) because nowhere in the body of `fn` is the outer  `def x` referenced.

The important thing to add here was the extra logic to the `AccessedFieldsVisitor` so that it transitively follows every method / field reference path possible. This change means we compute the `accessedFieldsMap` just once at the beginning and then the information it contains is sufficient for the cleaning, including the added step of nulling out outers in the hierarchy if they are never referenced.

andrewsmartin committed:
Check that closure is not serializable before cleaning.

andrewsmartin commented:
@johnynek I think I've addressed your comments, hope it's a bit easier to follow now. Some of the method names I felt were appropriate to change since they don't take a class / object parameter anymore (for example `outerClassesOf -> outerClasses`), but happy to change them if you prefer.

andrewsmartin commented:
Also the build is failing because of the breaking API change, should I bump the minor version too?

andrewsmartin commented:
@johnynek ping, are you fine with the changes?

johnynek commented:
Sorry I lost track of this. Looking now.

johnynek commented:
He 2.12 build is actually an error. Can you look at that?

For the others you can filter those errors. See the messages in the build and this example:

https://github.com/lightbend/migration-manager/wiki/Sbt-plugin#advanced-usage-filtering-binary-incompatibilities

As soon as we get the build to go green I am happy to merge.

andrewsmartin committed:
Make class reader use Option

johnynek commented:
Seems the builds all failed, each for seemingly different reasons. 2.10 needs to add a mima exclusion. 2.12 just straight-up fails. Not sure about 2.11.

johnynek commented:
If we could get this green, I'd still love to merge.

johnynek commented:
incorporated in #351
